muspy/base.py

```diff
@@ -186,17 +197,25 @@ def to_ordered_dict(self, skip_none: bool = True) -> OrderedDict:
                     continue
                 if isclass(attr_type) and issubclass(attr_type, Base):
                     ordered_dict[attr] = [
-                        v.to_ordered_dict(skip_none=skip_none) for v in value
+                        v.to_ordered_dict(skip_none=skip_none, copy=copy)
+                        for v in value
                     ]
-                else:
-                    ordered_dict[attr] = value
-            elif value is None:
+                    continue
+            if value is None:
                 if not skip_none:
                     ordered_dict[attr] = None
-            elif isclass(attr_type) and issubclass(attr_type, Base):
-                ordered_dict[attr] = value.to_ordered_dict(skip_none=skip_none)
+                continue
+            if isclass(attr_type) and issubclass(attr_type, Base):
+                ordered_dict[attr] = value.to_ordered_dict(
+                    skip_none=skip_none, copy=copy)
+                continue
+
+            if attr in self._list_attributes:
+                # List of non-MusPy objects
+                ordered_dict[attr] = deepcopy(value) if copy else list(value)
             else:
-                ordered_dict[attr] = value
+                # Single non-MusPy object
+                ordered_dict[attr] = deepcopy(value) if copy else value
```

Comment:
Note sure why we need `list(value)` for list attributes if `copy` is False. `list(value)` will create a copy of the list. I thought we should keep the original value no matter what it is.

Reply:
You can also think of it as:
```python
ordered_dict[attr] = [deepcopy(v) if copy else v for v in value]
```
I felt it would be more consistent this way. Basically, whenever we have a list attribute, we make a new list whether or not the values are MusPy objects. So if they are MusPy objects or primitives (immutable), everything is safe either way. If we have some other objects (which would be the case with annotations), then we have the question whether to copy or not, and that's where the caller gets to choose.

Reply:
It's different as a new list will always be created here, which is different from `list.extend`. I think we should try to keep the original list as how `list.extend` does.

```python
>>> import copy
>>> a = [1]
>>> b = [2]
>>> l = [[a], [b]]
>>> id(l), id(list(l)), id(copy.copy(l)), id(copy.deepcopy(l))
(2188910227016, 2188910245256, 2188910245256, 2188910245256)
>>> id(l[0]), id(list(l)[0]), id(copy.copy(l)[0]), id(copy.deepcopy(l)[0])
(2188910216904, 2188910216904, 2188910216904, 2188911381768)
```

We can see that `list(l)` create a new list similar to `copy.copy(l)`, while `list.extend` does not create a new list.

```python
>>> l.extend([0])
>>> id(l)
2188910227016
```

Reply:
Sure, I know this, but I'm not sure what `list.extend` has to do with this part of the code. My point was that when you do `music.to_ordered_dict()`, some of the lists in the resulting dict would be new lists while some of them would be shared with the original object, which might be confusing.

Anyway, let's keep this the way you find it logical.

Reply:
For me, `music.to_ordered_dict()` is a _view_ of the object and the value should point to the same data by default (like `np.reshape`). It should also be fast as `print` and `pretty_str` will call it.

Will any new list be created when `deepcopy=False`? I could miss something here. Let me know if so.

I hope `music.extend` works just like `list.extend` because I think it's more pythonic and that's someone who didn't read the code would expect (hopefully).

Reply:
[Here](https://github.com/salu133445/muspy/blob/7fba527085140c2b07e22df182595bd6097bb4ae/muspy/base.py#L167-L169), a new list is created for each list attribute whose type is a subclass of `Base`. This new list is by no means a view of the original list; also all the newly created `OrderedDicts` are also not views of the original objects. Unless *I* am missing something.